### ISSUE TITLE
Transformer's pad token has non-zero embedding

### ISSUE THREAD

**Issue by erip** [facebookresearch/fairseq#4633]
## 🐛 Bug

By default a transformer's token embedding layer has `padding_idx=1` indicating that the embedding of this token should be a 0 vector and should not contribute to loss. However, transformers trained using fairseq have non-zero pad token embeddings.

### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Train a model
2. Look at the token embedding weights

#### Code sample

The following fails with a transformer trained using fairseq:

```python
def check_zero_embedding(checkpoint_path, padding_idx=1):
    weights = torch.load(checkpoint_path)['model']['encoder.embed_tokens.weight']
    assert torch.all(weights[padding_idx] == 0)
```

alternatively, examine for WMT19 system:

```python
#!/usr/bin/env python3

import torch
import logging

# disable noisy fairseq logging...
logging.disable(1_000)

if __name__ == "__main__":
    en2de = torch.hub.load('pytorch/fairseq', 'transformer.wmt19.en-de.single_model')
    pad_embedding_weight = en2de.models[0].encoder.embed_tokens.weight[en2de.models[0].encoder.embed_tokens.padding_idx]
    assert torch.all(pad_embedding_weight == 0), f"Embedding for padding index should be 0, found {pad_embedding_weight}"

```

which fails:

```console
$ python failing_test.py
Using cache found in /Users/erip/.cache/torch/hub/pytorch_fairseq_main
Loading codes from /Users/erip/.cache/torch/pytorch_fairseq/81a0be5cbbf1c106320ef94681844d4594031c94c16b0475be11faa5a5120c48.63b093d59e7e0814ff799bb965ed4cbde30200b8c93a44bf8c1e5e98f5c54db3/bpecodes ...
Read 30000 codes from the codes file.
Traceback (most recent call last):
  File "/Users/erip/failing_test.py", line 11, in <module>
    assert torch.all(pad_embedding_weight == 0), f"Embedding for padding index should be 0, found {pad_embedding_weight}"
AssertionError: Embedding for padding index should be 0, found tensor([ 0.0109,  0.0018,  0.0024,  ...,  0.0170, -0.0071, -0.0250],
       grad_fn=<SelectBackward0>)
```

### Environment

 - fairseq Version (e.g., 1.0 or main): main
 - PyTorch Version (e.g., 1.0) 1.12
 - OS (e.g., Linux): n/a
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): ...
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by gmryu** [comment]
I believe the weight is randomly generated. At least, woud you confirm pad's embedding weight change after training?

Since [pad's embeds are set to zero in foward](https://github.com/facebookresearch/fairseq/blob/main/fairseq/models/transformer/transformer_encoder.py#L203-L211),
pad's embeddings should not contribute to loss.

**Comment by erip** [comment]
How do you expect the weights changed from the zero vector unless they were updated through gradient descent? I'm reporting this explicitly because it's happening in my network and I suspect it's causing issues with training. Even the native PyTorch embedding initializes the pad idx row in the embedding matrix w/ a zero-vector.

**Comment by gmryu** [comment]
@erip  You are correct. Pad's embedding weight was initialized with 0-vector here: https://github.com/facebookresearch/fairseq/blob/main/fairseq/models/transformer/transformer_base.py#L189-L193

Then if pad's embedding weight is updated during training (so this is why pretrained model has it?), this reported result implies the relationship with training loss.
